custom-metrics-stackdriver-adapter/pkg/apiserver/installer/emhandlers.go

```diff
@@ -0,0 +1,143 @@
+	context := a.group.Context
+
+	optionsExternalVersion := a.group.GroupVersion
+	if a.group.OptionsExternalVersion != nil {
+		optionsExternalVersion = *a.group.OptionsExternalVersion
+	}
+
+	mapping, err := a.restMapping()
+	if err != nil {
+		return err
+	}
+
+	fqKindToRegister, err := a.getResourceKind(a.group.DynamicStorage)
+	if err != nil {
+		return err
+	}
+
+	kind := fqKindToRegister.Kind
+
+	lister := a.group.DynamicStorage.(rest.Lister)
+	list := lister.NewList()
+	listGVKs, _, err := a.group.Typer.ObjectKinds(list)
+	if err != nil {
+		return err
+	}
+	versionedListPtr, err := a.group.Creater.New(a.group.GroupVersion.WithKind(listGVKs[0].Kind))
+	if err != nil {
+		return err
+	}
+	versionedList := indirectArbitraryPointer(versionedListPtr)
+
+	versionedListOptions, err := a.group.Creater.New(optionsExternalVersion.WithKind("ListOptions"))
+	if err != nil {
+		return err
+	}
+
+	ctxFn := func(req *http.Request) request.Context {
+		if ctx, ok := context.Get(req); ok {
+			return request.WithUserAgent(ctx, req.Header.Get("User-Agent"))
+		}
+		return request.WithUserAgent(request.NewContext(), req.Header.Get("User-Agent"))
+	}
+
+	scope := mapping.Scope
+	namespaceParam := ws.PathParameter(scope.ArgumentName(), scope.ParamDescription()).DataType("string")
+	nameParam := ws.PathParameter("name", "name of the described resource").DataType("string")
+
+	externalMetricParams := []*restful.Parameter{
+		namespaceParam,
+		nameParam,
+	}
+	externalMetricPath := scope.ParamName() + "/{" + scope.ArgumentName() + "}/{resource}"
+
+	mediaTypes, streamMediaTypes := negotiation.MediaTypesForSerializer(a.group.Serializer)
+	allMediaTypes := append(mediaTypes, streamMediaTypes...)
+	ws.Produces(allMediaTypes...)
+
+	reqScope := handlers.RequestScope{
+		ContextFunc:     ctxFn,
+		Serializer:      a.group.Serializer,
+		ParameterCodec:  a.group.ParameterCodec,
+		Creater:         a.group.Creater,
+		Convertor:       a.group.Convertor,
+		Copier:          a.group.Copier,
+		Typer:           a.group.Typer,
+		UnsafeConvertor: a.group.UnsafeConvertor,
+
+		// TODO: This seems wrong for cross-group subresources. It makes an assumption that a subresource and its parent are in the same group version. Revisit this.
+		Resource:    a.group.GroupVersion.WithResource("*"),
+		Subresource: "*",
+		Kind:        fqKindToRegister,
+
+		MetaGroupVersion: metav1.SchemeGroupVersion,
+	}
+	if a.group.MetaGroupVersion != nil {
+		reqScope.MetaGroupVersion = *a.group.MetaGroupVersion
+	}
+
+	doc := "list external metrics"
+	reqScope.Namer = MetricsNaming{
+		handlers.ContextBasedNaming{
+			GetContext:         ctxFn,
+			SelfLinker:         a.group.Linker,
+			ClusterScoped:      false,
+			SelfLinkPathPrefix: gpath.Join(a.prefix, scope.ParamName()) + "/",
+		},
+	}
+
+	externalMetricHandler := metrics.InstrumentRouteFunc("LIST", "external-metrics", "", "", restfulListResource(lister, nil, reqScope, false, a.minRequestTimeout))
```

Comment:
"external-metrics" is no longer part of path in the final version of API. Should it be used here?

Reply:
I think it's just an arbitrary name that doesn't correspond to anything in the path, similar to https://github.com/kubernetes-incubator/custom-metrics-apiserver/blob/master/pkg/apiserver/installer/installer.go#L297.